Model/Service/Index.php

```diff
@@ -0,0 +1,467 @@
+     * Create one if row does not exits
+     * Else set row to dirty
+     *
+     * @param ProductCollection $collection
+     * @param Store $store
+     */
+    public function handleProductChange(ProductCollection $collection, Store $store)
+    {
+        foreach ($collection as $product) {
+            $this->updateOrCreateDirtyEntity($product, $store);
+        }
+    }
+
+    /**
+     * @param Product $product
+     * @param Store $store
+     * @return ProductIndexInterface|null
+     */
+    public function updateOrCreateDirtyEntity(Product $product, Store $store)
+    {
+        $indexedProduct = $this->indexRepository->getByProductIdAndStoreId($product->getId(), $store->getId());
+        try {
+            if ($indexedProduct instanceof ProductIndexInterface) {
+                $indexedProduct->setIsDirty(true);
+                $indexedProduct->setUpdatedAt($this->magentoTimeZone->date());
+            } else {
+                /* @var Product $fullProduct */
+                /** @noinspection PhpParamsInspection */
+                $fullProduct = $this->loadMagentoProduct($product->getId(), $store->getId());
+                $indexedProduct = $this->indexBuilder->build($fullProduct, $store);
+                $indexedProduct->setIsDirty(false);
+            }
+            $this->indexRepository->save($indexedProduct);
+            return $indexedProduct;
+        } catch (\Exception $e) {
+            $this->logger->exception($e);
+            return null;
+        }
+    }
+
+    /**
+     * @param NostoIndexCollection $collection
+     * @param Store $store
+     * @throws NostoException
+     */
+    public function handleDirtyProducts(NostoIndexCollection $collection, Store $store)
+    {
+        $collection->setPageSize(self::PRODUCT_DATA_BATCH_SIZE);
+        $lastPage = $collection->getLastPageNumber();
+        $curPage = 1;
+        $totalDirty = 0;
+        do {
+            $collection->clear();
+            $collection->setCurPage($curPage);
+            foreach ($collection as $productIndex) {
+                if ($productIndex->getIsDirty() === NostoProductIndex::DB_VALUE_BOOLEAN_TRUE) {
+                    $this->rebuildDirtyProduct($productIndex);
+                    $totalDirty++;
+                }
+            }
+            $this->handleProductSync($collection, $store);
+            ++$curPage;
+        } while ($curPage <= $lastPage);
+        $this->logger->logWithMemoryConsumption(
+            sprintf(
+                'Rebuilt total of %d dirty products in %d batches',
+                $totalDirty,
+                $lastPage
+            )
+        );
+    }
+
+    /**
+     * Handles sync of product collection by sending it to Nosto through API
+     *
+     * @param NostoIndexCollection $collection
+     * @param Store $store
+     * @throws NostoException
+     */
+    public function handleProductSync(NostoIndexCollection $collection, Store $store): void
```

Comment:
This entire method needs to be refactored. I think that this is a violation of SRP and needs to be modularised. We should do this as the last change.